### PR TITLE
Upgrade to scalameta 2.0.0-M3

### LINKED ISSUE #169: Shorten fully qualified names from ExplicitReturnTypes.
ExplicitReturnTypes currently inserts fully qualified names, even if the short name is available in scope
```scala
// before
implicit val tt = liftedType
// after
implicit val tt: slick.api.TypedType[R] = liftedType
// would be nice
import slick.api.TypedType
implicit val tt: TypedType[R] = liftedType
```

### LINKED ISSUE #254: ExplicitReturnTypes doesn't fully qualify enough
After running ExplicitReturnTypes in the Akka repo I got the errors below. The full diff is available here https://gist.github.com/olafurpg/82b6fa8261ba2f1dc147dccd595098ab

```scala
[error] [1] /akka-actor/src/main/scala/akka/actor/ActorPath.scala:346:
[error]     type mismatch;
[error]      found   : java.lang.StringBuilder
[error]      required: scala.collection.mutable.StringBuilder
[error]         val sb: StringBuilder = buildToString(new JStringBuilder(length + 12), length, 0, _.toString)
[error]                                              ^
[error] [2] /akka-actor/src/main/scala/akka/actor/ActorPath.scala:347:
[error]     type mismatch;
[error]      found   : scala.collection.mutable.StringBuilder
[error]      required: java.lang.StringBuilder
[error]         appendUidFragment(sb).toString
[error]                           ^
[error] [3] /akka-actor/src/main/scala/akka/actor/ActorPath.scala:366:
[error]     type mismatch;
[error]      found   : java.lang.StringBuilder
[error]      required: scala.collection.mutable.StringBuilder
[error]         val sb: StringBuilder = buildToString(new JStringBuilder(length + 12), length, diff, _.toStringWithAddress(addr))
[error]                                              ^
[error] [4] /akka-actor/src/main/scala/akka/actor/ActorPath.scala:367:
[error]     type mismatch;
[error]      found   : scala.collection.mutable.StringBuilder
[error]      required: java.lang.StringBuilder
[error]         appendUidFragment(sb).toString
[error]                           ^
[error] [5] /akka-actor/src/main/scala/akka/actor/ActorSelection.scala:101:
[error]     type mismatch;
[error]      found   : java.lang.StringBuilder
[error]      required: scala.collection.mutable.StringBuilder
[error]         val builder: StringBuilder = new java.lang.StringBuilder()
[error]                                      ^
[error] [6] /akka-actor/src/main/scala/akka/actor/ActorSelection.scala:131:
[error]     type mismatch;
[error]      found   : java.lang.StringBuilder
[error]      required: scala.collection.mutable.StringBuilder
[error]         val builder: StringBuilder = new java.lang.StringBuilder()
[error]                                      ^
[error] [7] /akka-actor/src/main/scala/akka/actor/Address.scala:53:
[error]     type mismatch;
[error]      found   : java.lang.StringBuilder
[error]      required: scala.collection.mutable.StringBuilder
[error]         val sb: StringBuilder = (new java.lang.StringBuilder(protocol)).append("://").append(system)
[error]                                                                                             ^
[error] [8] /akka-actor/src/main/scala/akka/actor/FSM.scala:118:
[error]     recursive value → needs type
[error]       val `→`: akka.actor.FSM. → .type = `->`
[error]                                ^
[error] [9] /akka-actor/src/main/scala/akka/dispatch/CachingConfig.scala:142:
[error]     type mismatch;
[error]      found   : java.util.List[java.lang.Boolean]
[error]      required: java.util.List[scala.Boolean]
[error]       def getBooleanList(path: String): java.util.List[Boolean] = config.getBooleanList(path)
[error]                                                                                        ^
[error] [10] /akka-actor/src/main/scala/akka/dispatch/CachingConfig.scala:148:
[error]      type mismatch;
[error]       found   : java.util.List[java.lang.Long]
[error]       required: java.util.List[scala.Long]
[error]        def getLongList(path: String): java.util.List[Long] = config.getLongList(path)
[error]                                                                                ^
[error] [11] /akka-actor/src/main/scala/akka/dispatch/CachingConfig.scala:150:
[error]      type mismatch;
[error]       found   : java.util.List[java.lang.Double]
[error]       required: java.util.List[scala.Double]
[error]        def getDoubleList(path: String): java.util.List[Double] = config.getDoubleList(path)
[error]                                                                                      ^
[error] [12] /akka-actor/src/main/scala/akka/dispatch/CachingConfig.scala:160:
[error]      type mismatch;
[error]       found   : java.util.List[java.lang.Long]
[error]       required: java.util.List[scala.Long]
[error]        def getBytesList(path: String): java.util.List[Long] = config.getBytesList(path)
[error]                                                                                  ^
[error] [13] /akka-actor/src/main/scala/akka/dispatch/CachingConfig.scala:162:
[error]      type mismatch;
[error]       found   : java.util.List[java.lang.Long]
[error]       required: java.util.List[scala.Long]
[error]        def getMillisecondsList(path: String): java.util.List[Long] = config.getDurationList(path, TimeUnit.MILLISECONDS)
[error]                                                                                            ^
[error] [14] /akka-actor/src/main/scala/akka/dispatch/CachingConfig.scala:164:
[error]      type mismatch;
[error]       found   : java.util.List[java.lang.Long]
[error]       required: java.util.List[scala.Long]
[error]        def getNanosecondsList(path: String): java.util.List[Long] = config.getDurationList(path, TimeUnit.NANOSECONDS)
[error]                                                                                           ^
[error] [15] /akka-actor/src/main/scala/akka/dispatch/CachingConfig.scala:178:
[error]      type mismatch;
[error]       found   : java.util.List[java.lang.Long]
[error]       required: java.util.List[scala.Long]
[error]        def getDurationList(path: String, unit: TimeUnit): java.util.List[Long] = config.getDurationList(path, unit)
[error]                                                                                                        ^
[error] [16] /akka-actor/src/main/scala/akka/event/Logging.scala:1285:
[error]      type mismatch;
[error]       found   : java.lang.StringBuilder
[error]       required: scala.collection.mutable.StringBuilder
[error]          val sb: StringBuilder = new java.lang.StringBuilder(64)
[error]                                  ^
```
It seems ExplicitReturnTypes must either do a smarter job of finding the correct name of those types or fully qualify more.

In the particular case of ActorPath.scala, there is a top-level import 

```
import java.lang.{ StringBuilder ⇒ JStringBuilder }
```

### PR DESCRIPTION
**Pull request by olafurpg**
The biggest change besides renames is the ability to query resolved names from Denotation.signature, this required rethinking ExplicitReturnTypes.

- experimental explicitReturnTypes.unsafeShortenNames, fixes #169 
- insert fully qualified names by default, fixes #254. Note, to 100% fix that problem the rewrite should insert _root_ as well, since there can be conflicts with names in scope.

### CODE REVIEW

File: scalafix-testkit/src/main/scala/scalafix/testkit/DiffTest.scala [hunk 1]
```diff
@@ -11,7 +11,7 @@ import org.scalatest.exceptions.TestFailedException
 case class DiffTest(
     filename: RelativePath,
     original: Input,
-    attributes: Attributes,
+    attributes: Document,
```
Comment (xeno-by) [inline scalafix-testkit/src/main/scala/scalafix/testkit/DiffTest.scala]: document?

File: scalafix-testkit/src/main/scala/scalafix/testkit/DiffTest.scala [hunk 2]
```diff
@@ -25,7 +25,7 @@ object DiffTest {
   private def stripPrefix(str: String) = PrefixRegex.replaceFirstIn(str, "")
 
   def fromSemanticCtx(sctx: SemanticCtx): Seq[DiffTest] =
-    sctx.entries.map { attributes =>
+    sctx.documents.map { attributes =>
```
Comment (xeno-by) [inline scalafix-testkit/src/main/scala/scalafix/testkit/DiffTest.scala]: document?

File: scalafix-tests/unit/src/test/scala/scalafix/tests/BaseSemanticTest.scala [hunk 3]
```diff
@@ -10,14 +10,14 @@ abstract class BaseSemanticTest(filename: String)
     extends FunSuite
     with BeforeAndAfterAll {
   private var _db: SemanticCtx = _
-  private var _attrs: Attributes = _
+  private var _docs: Document = _
```
Comment (xeno-by) [inline scalafix-tests/unit/src/test/scala/scalafix/tests/BaseSemanticTest.scala]: _doc

File: scalafix-core/shared/src/main/scala/scalafix/syntax/package.scala [hunk 4]
```diff
@@ -33,7 +33,7 @@ package object syntax {
   implicit class XtensionSymbol(symbol: Symbol) {
     def normalized: Symbol = SymbolOps.normalize(symbol)
   }
-  implicit class XtensionAttributes(attributes: Attributes) {
+  implicit class XtensionDocument(attributes: Document) {
```
Comment (xeno-by) [inline scalafix-core/shared/src/main/scala/scalafix/syntax/package.scala]: document?

**Comment by olafurpg** [comment]
I addressed your comments in #321

**Pull request merged**